packages/attune/src/server/repo/index/sign.rs

```diff
@@ -150,526 +173,532 @@ pub async fn handler(
     }
 
     // Save the new state to the database.
-    // The old hash values are for cleanup after uploading new content.
-    let old_hashes = match req.change.action {
-        PackageChangeAction::Add { .. } => {
-            // First, we update-or-create the Release. Remember, it's possible that no
-            // package has ever been added to this distribution, so the Release may not
-            // exist.
-            let release_id = match sqlx::query!(r#"
+    let previous_by_hash_indexes = match req.change.action {
+        PackageChangeAction::Add { .. } => add_package_to_db(tx, tenant_id, req, &result).await,
+        PackageChangeAction::Remove {
+            ref name,
+            ref version,
+            ref architecture,
+        } => Some(
+            remove_package_from_db(tx, tenant_id, req, &result, name, version, architecture).await,
+        ),
+    };
+
+    Ok((result, previous_by_hash_indexes))
+}
+
+#[derive(Debug)]
+struct PreviousByHashIndexes {
+    md5sum: String,
+    sha1sum: String,
+    sha256sum: String,
+}
+
+async fn add_package_to_db(
+    tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    tenant_id: &TenantID,
+    req: &SignIndexRequest,
+    update: &PackageChangeResult,
+) -> Option<PreviousByHashIndexes> {
+    // First, we update-or-create the Release. Remember, it's possible that no
+    // package has ever been added to this distribution, so the Release may not
+    // exist.
```

Comment:
[optional] while we're touching this already, consider updating this and other upserts in this diff to just use a `INSERT INTO ... ON CONFLICT DO UPDATE ... RETURNING` query, which should cover all 3 cases in one go.

Reply:
For this _particular_ upsert, I'm not doing `ON CONFLICT DO UPDATE` because:
1. I think it's actually extremely rare that releases will be updated, and it mildly annoys me that an `ON CONFLICT DO UPDATE` will still increment the primary key sequence.
2. I don't think doing multiple statements in a transaction adds that much overhead.

The other queries in this function have similar weirdness:
1. Components don't actually get added that much.
2. `Packages` indexes need their old hashes, but only conditionally.

We could use `MERGE`, but I'm not familiar with its footguns. And it's not _that bad_ to burn serial IDs, but also it's not that bad to add a little bit of statement overhead either. I'm going to keep this currently configuration for now until I am compelled otherwise.